lib/containers/runtime.pm

```diff
@@ -131,18 +222,178 @@ sub remove_image {
+    my $keep_container = $args{keep_container} ? ''                   : '--rm';
+    my $params         = sprintf qq(%s %s %s), $keep_container, $mode, $name;
+    my $ret            = $self->_rt_script_run(sprintf qq(run %s %s %s), $params, $image_name, $remote, timeout => $args{timeout});
+    record_info "Remote run on $image_name", "options $keep_container $mode $name $image_name $remote";
+    return $ret;
+}
+
+package containers::runtime::podman;
+use Mojo::Base 'containers::runtime';
+use testapi;
+use containers::utils qw(registry_url);
+use utils qw(systemctl file_content_replace);
+has runtime => "podman";
+
+sub setup_registry {
+    my ($self) = shift;
+    my $registry = registry_url();
+
+    assert_script_run "curl " . data_url('containers/registries.conf') . " -o /etc/containers/registries.conf";
+    assert_script_run "chmod 644 /etc/containers/registries.conf";
+    file_content_replace("/etc/containers/registries.conf", REGISTRY => $registry);
+}
+
+
+package containers::runtime::buildah;
+use Mojo::Base 'containers::runtime';
+use testapi;
+use containers::utils qw(registry_url);
+use utils qw(systemctl file_content_replace);
+has runtime => "buildah";
+
+sub cleanup_system_host {
+    my ($self, $assert) = @_;
+    $assert // 1;
+    $self->_rt_assert_script_run("rm --all");
+    $self->_rt_assert_script_run("rmi --all --force");
+
+    if ($assert) {
+        assert_equals(0, scalar @{$self->enum_containers()}, "containers have not been removed");
+        assert_equals(0, scalar @{$self->enum_images()},     "images have not been removed");
+    }
+}
+
+sub get_images_by_repo_name {
+    my ($self) = @_;
+    my $repo_images;
+
+    $repo_images = $self->_rt_script_output("images --format '{{.Name}}'");
+    my @images = split /[\n\t ]/, $repo_images;
+    return \@images;
+}
+
+sub up {
+    my ($self, $image_name, %args) = @_;
+    die 'image name or id is required' unless $image_name;
+    my $remote = $args{cmd} ? "-- $args{cmd}" : '';
+    $image_name = $args{name} ? "\$(buildah from $args{name})" : '$image_name';
+    my $ret = $self->_rt_script_run(sprintf qq(run %s %s %s), $image_name, $remote, timeout => $args{timeout});
+    record_info "Remote run on $image_name", "options $image_name $remote";
+    return $ret;
+}
+
+sub pull {
+    my ($self, $image_name, %args) = @_;
+    if (my $rc = $self->_rt_script_run("images | grep '$image_name'") == 0) {
+        return;
+    }
+    # At least on publiccloud, this image pull can take long and occasinally fails due to network issues
+    return $self->_rt_script_run("pull $image_name", timeout => $args{timeout} // 300);
+}
+
+sub create_container {
+    my ($self, $image, $name, $args) = @_;
+    die 'wrong number of arguments' if @_ < 3;
+    my $container = $self->_rt_script_output("from $image 2>/dev/null");
+    record_info 'Container', qq[Testing:\nContainer "$container" based on image "$image"];
+}
```

Comment:
where is the container created here in this function?

Reply:
L376

Reply:
ok, `buildah from <image>`, just out of curiosity, why redirecting stderr?

Reply:
https://github.com/os-autoinst/os-autoinst-distri-opensuse/commit/0c45b7fcf3c7a6c30d8ba7f59ef2025e3fe58adc